### ISSUE TITLE
Remove stopped transceivers from getTransceivers()

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-pc#2092]
Why keep them around when indices don't necessarily match the SDP anyway?

**Comment by nils-ohlmeier** [comment]
Isn't Bernards comment https://github.com/w3c/webrtc-pc/issues/1975#issuecomment-417442346 still true: that stopped Transceivers does not mean the other side isn't using that m-section any more?

**Comment by alvestrand** [comment]
Indices in the transceiver list don't match the SDP m-line index in general (it does at the time of initial generation of the SDP offer, but given that the indices are in the order in which transcevers were added to the list, transceivers that were created prior to generating an answer won't be slotted into corresponding m-line indexes, and the rules for recycling m-sections also mean that the indices don't match.

Until @jan-ivar updated the transceiver list getter to require time-of-addition order (so that it was stable), it was legal to return the list in m-line-index order, presumably with loose transceivers at the end; this meant that getTransceivers() called before and after a negotiation could give back different orders even if no transceivers were stopped; this is now illegal (and needs test).

I'm happy to throw stopped transceivers out of the list, since I think referring to them by list index is foolish anyway; note that if the app has stored references to them, that doesn't mean the objects disappear.

Aboba's comment refers to stopping the receiver's track, not stopping the transceiver (which stops both sending and receiving, and triggers negotiation-needed).

**Comment by jan-ivar** [comment]
@alvestrand Without deterministic transceiver order, how do you correlate tracks on initial offer?

I cover this use-case in my [Exploring RTCRtpTransceiver](https://blog.mozilla.org/webrtc/rtcrtptransceiver-explored/) blog post, under the "Well-defined transceiver order" sub-heading, which I can't link to (Thanks WordPress!)

**Comment by henbos** [comment]
@jan-ivar You look at the mid, at what stage do you ever have to care about the index of a transceiver? It's enough to have the transceiver references. That gives you both mids and tracks.

**Comment by henbos** [comment]
Unless I'm missing something, even your own example doesn't care about the indices, it does this:
```
    let transceivers = [
      pc1.addTransceiver(camStream.getVideoTracks()[0]),
      pc1.addTransceiver(camStream.getAudioTracks()[0]),
      pc1.addTransceiver(noise)
    ];
```
And then wires pc.ontrack in such a way that I don't think the order of things firing or the transceiver indices matters.

**Comment by jan-ivar** [comment]
From [Exploring RTCRtpTransceiver](https://blog.mozilla.org/webrtc/rtcrtptransceiver-explored/) blog post, under the "Well-defined transceiver order" sub-heading:

> ### Well-defined transceiver order.
> Something I overlooked last year is that `transceiver.mid` is `null` before `setLocalDescription`. We avoided that problem above by establishing the connection ahead of sending anything, but this makes `mid` useless for correlating in the initial negotiation.
> 
> The good news here since last year is that `pc.getTransceivers()` order is now well-defined! Transceivers always appear in creation-order, whether introduced by you with `addTransceiver`, or in m-line order from `setRemoteDescription`. That m-line order matches the other side’s transceiver order, at least initially.
> 
> With some care, this means we can correlate tracks using transceiver order from the initial offer itself. Here’s an example—without check-boxes this time—that does that. We’re also introducing a microphone into the picture: 
> ```js
> let camStream;
> 
> (async () => {
>   try {
>     const [pc1, pc2] = localPeerConnectionLoop();
>     camStream = await navigator.mediaDevices.getUserMedia({video: true, audio: true});
>     let transceivers = [
>       pc1.addTransceiver(camStream.getVideoTracks()[0], {streams: [camStream]}),
>       pc1.addTransceiver(camStream.getAudioTracks()[0], {streams: [camStream]}),
>       pc1.addTransceiver(whiteNoise())
>     ];
> 
>     pc2.ontrack = ({transceiver, streams: [stream]}) => {
>       if (transceiver.mid == pc2.getTransceivers()[2].mid) {
>         videoB.srcObject = new MediaStream([transceiver.receiver.track]);
>       } else {
>         videoA.srcObject = stream;
>       }
>     }
>   } catch (e) {
>     console.log(e)
>   }
> })();
> ```
> In the “Result” tab, you’ll see the camera on the left again. You can mute the audio in that video element as well.
> 
> This time in `pc2.ontrack`, we don’t cheat by looking at the other side’s `transceivers`. We only look at our own `pc2.getTransceivers()` which is **guaranteed to be in the same order here**.
> 
> Specifically, we look if this is the third transceiver (`pc2.getTransceivers()[2]`), and if so, put it on the right, otherwise left. We also use the `streams` argument to intuitively group the camera and microphone tracks together. Since the third track didn’t have a stream in this case, we could have keyed off of that difference instead. There may be several ways to correlate at times: by transceiver order, by stream, or by out-of-band `mid`.
> 
> If you’re wondering how we can access the third transceiver already in the first `ontrack`, the API guarantees that `setRemoteDescription` is effectively done by the time the `track` events all fire. All transceivers are there; all tracks are in their streams.

**Comment by henbos** [comment]
OK. But shouldn't this be viewed as a special case where we have not ruined the order (e.g. by adding transceivers before hand)? Removing transceivers does not break this use case. I think the beauty is that "ontrack" will fire in the same order, not that the index is the same (because it is not always the same).

**Comment by henbos** [comment]
@jan-ivar Let's continue the discussion over at the public-webrtc@ mailing list entitled "Proposal: Remove stopped transceivers to prevent resource exhaustion" if you have more comments

**Comment by jan-ivar** [comment]
I think I can finish up this sub-thread here. I was responding to @alvestrand's *"referring to them by list index is foolish"*.

> I think the beauty is that "ontrack" will fire in the same order, not that the index

Same order as what? JS needs a reference. `getTransceivers()` used to be in no order whatsoever. Not sure we want to go back to that. Also, keeping track of track events (no pun intended) is harder in practice than it might seem (e.g. developer: *"Is the first track of the next bunch, or a (potentially new) last track of the previous bunch?"*). Developers would need to keep track of signaling state to separate one group of events from the next. What could go wrong?

For other examples of relying on transceiver order, see [web-platform-tests](https://github.com/web-platform-tests/wpt/blob/master/webrtc/RTCRtpTransceiver.https.html#L1033).

> Removing transceivers does not break this use case.

Agree, just wanted to make sure this didn't devolve into removing transceiver order.

The somewhat related problem is: removing a stopped transceiver other than the last one, changes the indices of all/any transceivers after it by 1, but we already have that problem with `getSenders` and `getReceivers`.

**Comment by aboba** [comment]
@henbos @jan-ivar  Would it be helpful to discuss this issue at the March Interim?  If so, can you prepare slide(s) for that discussion?

**Comment by henbos** [comment]
I've made a slide for this.

**Comment by jan-ivar** [comment]
I said yesterday in review I was having [doubts about this](https://github.com/w3c/webrtc-pc/pull/2102#discussion_r264942039). I've slept on it, and I see a problem: stats.

Stats closely [model internal object lifetimes](http://w3c.github.io/webrtc-pc/#introduction-0), so this isn't just for people who care about stats. Today, our model is that the majority of [monitored objects](http://w3c.github.io/webrtc-pc/#dfn-monitored-object) live for the lifetime of the peer connection, beyond `pc.close()` so apps can query what happened posthumously. Only objects actively *replaced*—TrackAttachmentStats by *replaceTrack*, and CandidateStats by *iceRestart*—get removed and fired in [statsended](http://w3c.github.io/webrtc-pc/#event-statsended), to curtail bloat over time (sound familiar?)

So...

Would related sender, receiver, rtp and candidate stats disappear from `pc.getStats()` on *stop()* now? On SRD(offer) with port=0? That might be hard to foresee. On *close()*? That might fire events after *close()*, something we've avoided thus far. What if I keep a reference to a *sender* and call `sender.getStats()` on it? Does it succeed?

In short, the lifetime question is harder than removing stopped transceiver from *getTransceivers()*. We have to specify when related objects become garbage collectable, in a way that doesn't let JS detect garbage collection.

If we *don't* remove the stats, we arguably haven't fully solved the transceiver bloat problem.

The silver lining here is that stats has already solved this once, just not this way.

**Comment by henbos** [comment]
getStats() should only return stats for objects that are "live". If the underlying object that the stats represent is removed, "onstatsended" should fire. This should happen to RTP stream stats already, and if that's not explicit then we should file a spec bug. So yes, if a transceiver is stopped I think it follows that we should fire "onstatsended" for the sender/receiver/rtp stats.

sender/receiver.getStats() is already defined as "gather stats for all RTP streams that are sent/received by the sender/receiver". Which means this already today, without the proposed change, returns an empty set before you've negotiated, because the [lifetime of RTP stream stats](https://w3c.github.io/webrtc-stats/#the-rtp-statistics-hierarchy) start when the first RTP packet is sent/received.

I think this needs to be fixed to include "sender" and "receiver" stats too in the case of calling getStats() before you have RTP streams because that makes more sense, but my point is that this is already a thing. Even if transceiver.stop() doesn't remove the transceiver, it DOES remove the RTP streams. This problem is tangential, and not currently a blocking issue.

And no, we should not decide whether or not to keep objects around based on GC. It is based on whether or not they are attached to the PC.

**Comment by henbos** [comment]
Sorry, the spec also says "RTP monitored objects are not deleted."

This is surprising to me :O
